test/src/lib_state_machine_test.cpp

```diff
@@ -0,0 +1,239 @@
+static void stage2Run() {}
+static void stage2Exit() {GLOBAL_CTX.flag_s2 = true;}
+
+static void stage3Run(){}
+static void stage3Entry(){GLOBAL_CTX.flag_s3 = true;}
+
+static void stage4Run() {}
+
+// The fixture for testing class Foo.
+class LibStateMachineTest : public ::testing::Test {
+ protected:
+  // You can remove any or all of the following functions if its body
+  // is empty.
+
+  LibStateMachineTest() {
+     // You can do set-up work for each test here.
+  }
+
+  ~LibStateMachineTest() override {
+     // You can do clean-up work that doesn't throw exceptions here.
+  }
+
+  // If the constructor and destructor are not enough for setting up
+  // and cleaning up each test, you can define the following methods:
+
+  void SetUp() override {
+     // Code here will be called immediately after the constructor (right
+     // before each test).
+     GLOBAL_CTX.count_s1 = 0;
+     GLOBAL_CTX.flag_s2 = false;
+     GLOBAL_CTX.flag_s3 = false;
+  }
+
+  void TearDown() override {
+     // Code here will be called immediately after each test (right
+     // before the destructor).
+  }
+
+  // Objects declared here can be used by all tests in the test suite for Foo.
+};
+
+// Although lib_sm_run should be run in a for(;;) loop, while writing unit tests for our state machines
+// We should seperate out each call to run() to more granularly test what's going on.
+TEST_F(LibStateMachineTest, BasicPump)
+{
+  // Check Initialization.
+  lib_sm_init(&GLOBAL_CTX.sm_ctx, &states[STAGE_1], checkTransitions);
+  EXPECT_TRUE(GLOBAL_CTX.sm_ctx.current_state == &states[STAGE_1]);
+  EXPECT_TRUE((strcmp(lib_sm_get_current_state_name(&GLOBAL_CTX.sm_ctx), "Stage1") == 0));
+
+  // Run 1 count_s1 == 1
+  lib_sm_run(&GLOBAL_CTX.sm_ctx);
+  EXPECT_TRUE(GLOBAL_CTX.sm_ctx.current_state == &states[STAGE_1]);
+  EXPECT_TRUE((strcmp(lib_sm_get_current_state_name(&GLOBAL_CTX.sm_ctx), "Stage1") == 0));
+
+  // Run 2 count_s1 == 2
+  lib_sm_run(&GLOBAL_CTX.sm_ctx);
+  EXPECT_TRUE(GLOBAL_CTX.sm_ctx.current_state == &states[STAGE_2]);
+  EXPECT_TRUE((strcmp(lib_sm_get_current_state_name(&GLOBAL_CTX.sm_ctx), "Stage2") == 0));
+
+  // Run 3
+  lib_sm_run(&GLOBAL_CTX.sm_ctx);
+  EXPECT_TRUE(GLOBAL_CTX.flag_s2 == true);
+  EXPECT_TRUE(GLOBAL_CTX.flag_s3 == true);
+  EXPECT_TRUE(GLOBAL_CTX.sm_ctx.current_state == &states[STAGE_3]);
+  EXPECT_TRUE((strcmp(lib_sm_get_current_state_name(&GLOBAL_CTX.sm_ctx), "Stage3") == 0));
+
+  // Run 4
+  lib_sm_run(&GLOBAL_CTX.sm_ctx);
+  EXPECT_TRUE(GLOBAL_CTX.sm_ctx.current_state == &states[STAGE_4]);
+  EXPECT_TRUE((strcmp(lib_sm_get_current_state_name(&GLOBAL_CTX.sm_ctx), "Stage4") == 0));
+
+  // Run 5
+  lib_sm_run(&GLOBAL_CTX.sm_ctx);
+  EXPECT_TRUE(GLOBAL_CTX.sm_ctx.current_state == &states[STAGE_1]);
+  EXPECT_TRUE((strcmp(lib_sm_get_current_state_name(&GLOBAL_CTX.sm_ctx), "Stage1") == 0));
+}
+
+// TODO - re-enable these tests once we have re-enabled ASSERTs. This test expects to crash,
```

Comment:
Do we want to re-enabel asserts, or should we continue going with improved error handling? I feel like it is a chore to refactor the code a bit, but also offers better information to an integrator rather than crashing the system.

Reply:
I think the goal should be to have bm_common and bm_core, as much as possible, return errors, or set some error notifications if not in a context where we can return, so that the code outside bm_core can decide what to do in case of those errors.

In our app code that then uses bm_core, we probably still want to assert in most of these cases, but we should do it outside the library if that's feasible.

These are also changes we don't have to make right now — we can queue them up for future work.

That's my "strong opinion loosely held" — if someone has a good argument for keeping the asserts inside the lib, conditionally enabled, I'm open to that too.

Reply:
I agree, I'll create a ticket to convert asserts to reporting errors.

Reply:
ticket created https://app.shortcut.com/sofar/iteration/207310/fw-2024-21-hello-world-ported?group_by=workflow_state_id